pnpm/src/switchCliVersion.ts

```diff
@@ -0,0 +1,53 @@
+import fs from 'fs'
+import path from 'path'
+import { type Config } from '@pnpm/config'
+import { globalWarn } from '@pnpm/logger'
+import { detectIfCurrentPkgIsExecutable, packageManager } from '@pnpm/cli-meta'
+import { prependDirsToPath } from '@pnpm/env.path'
+import spawn from 'cross-spawn'
+import semver from 'semver'
+import { pnpmCmds } from './cmd'
+
+export async function switchCliVersion (config: Config): Promise<void> {
+  const pm = config.wantedPackageManager
+  if (pm == null || pm.name !== 'pnpm' || pm.version == null || pm.version === packageManager.version) return
+  if (!semver.valid(pm.version)) {
+    globalWarn(`Cannot switch to pnpm@${pm.version}`)
```

Comment:
I don't think this message is clear enough. The user could have mistakenly specify a version range and then be confused.

```suggestion
    globalWarn(`Cannot switch to pnpm@${pm.version}: ${pm.version} is not a valid version`)
```